test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -79,167 +87,330 @@ func (s *snapshottableTestSuite) SkipRedundantSuite(driver TestDriver, pattern t
-		volumeSnapshotRef := snapshotContentSpec["volumeSnapshotRef"].(map[string]interface{})
+	})
+
+	ginkgo.It("should not delete snapshot with retain policy [Feature:VolumeSnapshotDataSource]", func() {
+		l := &snapshottableLocal{}
 
-		// Check SnapshotContent properties
-		ginkgo.By("checking the SnapshotContent")
-		framework.ExpectEqual(snapshotContentSpec["volumeSnapshotClassName"], vsc.GetName())
-		framework.ExpectEqual(volumeSnapshotRef["name"], snapshot.GetName())
-		framework.ExpectEqual(volumeSnapshotRef["namespace"], snapshot.GetNamespace())
+		init(l)
+		defer cleanup(l)
+
+		TestSnapshottable(l, SnapshotClassTest{
+			DeletionPolicy: "Retain",
+		})
+		TestSnapshotDeleted(l, SnapshotClassTest{
+			DeletionPolicy: "Retain",
+		})
 	})
 }
 
+// snapshottableLocal is used to keep the current state of a snapshottable
+// test, associated objects, and cleanup steps.
+type snapshottableLocal struct {
+	config        *PerTestConfig
+	driverCleanup func()
+	cleanupSteps  []func()
+
+	cs        clientset.Interface
+	dc        dynamic.Interface
+	pvc       *v1.PersistentVolumeClaim
+	sc        *storagev1.StorageClass
+	pod       *v1.Pod
+	vsc       *unstructured.Unstructured
+	vs        *unstructured.Unstructured
+	vscontent *unstructured.Unstructured
+}
+
+// SnapshotClassTest represents parameters to be used by snapshot tests.
+// Not all parameters are used by all tests.
+type SnapshotClassTest struct {
+	DeletionPolicy string
+}
+
+// TestSnapshottable tests volume snapshots based on a given SnapshotClassTest
+func TestSnapshottable(l *snapshottableLocal, sct SnapshotClassTest) {
+	var err error
+
+	ginkgo.By("creating a SnapshotClass")
+	l.vsc.Object["deletionPolicy"] = sct.DeletionPolicy
+	l.vsc, err = l.dc.Resource(SnapshotClassGVR).Create(context.TODO(), l.vsc, metav1.CreateOptions{})
+	framework.ExpectNoError(err)
+	l.cleanupSteps = append(l.cleanupSteps, func() {
+		framework.Logf("deleting SnapshotClass %s", l.vsc.GetName())
+		l.dc.Resource(SnapshotClassGVR).Delete(context.TODO(), l.vsc.GetName(), metav1.DeleteOptions{})
+	})
+	l.vsc, err = l.dc.Resource(SnapshotClassGVR).Namespace(l.vsc.GetNamespace()).Get(context.TODO(), l.vsc.GetName(), metav1.GetOptions{})
+	framework.ExpectNoError(err)
+
+	ginkgo.By("creating a snapshot")
+	l.vs = getSnapshot(l.pvc.Name, l.pvc.Namespace, l.vsc.GetName())
+
+	l.vs, err = l.dc.Resource(SnapshotGVR).Namespace(l.vs.GetNamespace()).Create(context.TODO(), l.vs, metav1.CreateOptions{})
+	framework.ExpectNoError(err)
+	l.cleanupSteps = append(l.cleanupSteps, func() {
+		framework.Logf("deleting snapshot %q/%q", l.vs.GetNamespace(), l.vs.GetName())
+		// typically this snapshot has already been deleted
+		err = l.dc.Resource(SnapshotGVR).Namespace(l.vs.GetNamespace()).Delete(context.TODO(), l.vs.GetName(), metav1.DeleteOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			framework.Failf("Error deleting snapshot %q. Error: %v", l.pvc.Name, err)
+		}
+	})
```

Comment:
I think we need not add delete snapshot to the l.cleanupsteps. this can be a simple defer since TestSnapshottable() is standalone. Keeps it uniform with the defer model in retain test.

Reply:
Moved to be defer